cpp/src/io/shp/polygon_shapefile_reader.cu

```diff
@@ -0,0 +1,244 @@
+            LinearRingFromPolygon(*((OGRPolygon *) poShape),aPointX, aPointY, aPartSize );
+	else
+           CUDF_EXPECTS(0, "must be polygonal geometry." );    
+    }
+ 
+     /*
+     * Read a GDALDataset layer (corresponding to a shapefile) into five vectors
+     *
+     * layer: OGRLayer layer holding polygon data
+     * g_len_v: vector of group lengths, i.e., numbers of features/polygons (should be 1 for a single layer/g_len_v)
+     * f_len_v: vector of feature lengths, i.e., numbers of rings in features/polygons
+     * r_len_v: vector of ring lengths, i.e., numbers of vertices in rings
+     * x_v: vector of x coordiantes of vertices
+     * y_v: vector of y coordiantes of vertices
+     * returns number of features/polygons
+     
+    */
+    
+    int ReadLayer(const OGRLayerH layer,std::vector<int>& g_len_v,std::vector<int>&f_len_v,
+         std::vector<int>& r_len_v,std::vector<double>& x_v, std::vector<double>& y_v)         
+    {
+        int num_feature=0;
+        OGR_L_ResetReading(layer );
+        OGRFeatureH hFeat;
+        while( (hFeat = OGR_L_GetNextFeature( layer )) != NULL )
+        {
+            OGRGeometry *poShape=(OGRGeometry *)OGR_F_GetGeometryRef( hFeat );
+            CUDF_EXPECTS(poShape!=NULL,"Invalid Shape");
+            
+            std::vector<double> aPointX;
+            std::vector<double> aPointY;
+            std::vector<int> aPartSize;
+ 	    PolygonFromGeometry( poShape, aPointX, aPointY, aPartSize );
+ 		
+            x_v.insert(x_v.end(),aPointX.begin(),aPointX.end());
+ 	    y_v.insert(y_v.end(),aPointY.begin(),aPointY.end());
+            r_len_v.insert(r_len_v.end(),aPartSize.begin(),aPartSize.end());
+ 	    f_len_v.push_back(aPartSize.size());
+ 	    OGR_F_Destroy( hFeat );
+ 	    num_feature++;
+        }
+        g_len_v.push_back(num_feature);
+        return num_feature;
+    }
+}
+
+namespace cuspatial
+{
+    /*
+    * Read a polygon shapefile and fill in a polygons structure
+    * ToDo: read associated relational data into a CUDF Table 
+    *
+    * filename: ESRI shapefile name (wtih .shp extension
+    * pm: structure polygons (fixed to double type) to hold polygon data
+    
+    * Note: only the first layer is read - shapefiles have only one layer in GDALDataset model    
+    */
+
+    void polygon_from_shapefile(const char *filename, struct polygons<double>& pm)
+    {
+        std::vector<int> g_len_v,f_len_v,r_len_v;
+        std::vector<double> x_v, y_v;
+        GDALAllRegister();
+        
+        GDALDatasetH hDS = GDALOpenEx( filename, GDAL_OF_VECTOR, NULL, NULL, NULL );
+        CUDF_EXPECTS(hDS!=NULL,"Failed to open ESRI Shapefile dataset");		    
+        OGRLayerH hLayer = GDALDatasetGetLayer( hDS,0 );
+        CUDF_EXPECTS(hLayer!=NULL,"Failed to open the first layer");
+        int num_f=ReadLayer(hLayer,g_len_v,f_len_v,r_len_v,x_v,y_v);
+        CUDF_EXPECTS(num_f>0,"Shapefile must have at lest one polygon");
+        
+        pm.num_group=g_len_v.size();
```

Comment:
Give `polygons` a constructor so you don't have to manually allocate and set every member.

Reply:
And a destructor.

Reply:
It is designed to be a struct. memset does the job, just like gdf_column. 
I want to avoid changes to existing code that has already been used previous versions. A more thorough review/redesign/re-implementation can be scheduled in future.